### PR TITLE
row_indexes for IJMatrixGetValues

### PR DESCRIPTION
**Pull request by waynemitchell**
Add option to pass row_indexes to the IJMatrixGetValues routines. This aligns with existing implementation for the SetValues routines and is necessary for upcoming functionality in the recmat branch.

### CODE REVIEW

File: src/IJ_mv/HYPRE_IJMatrix.c [hunk 1]
```diff
@@ -814,7 +867,133 @@ HYPRE_IJMatrixGetValues( HYPRE_IJMatrix matrix,
...
+      return hypre_error_flag;
+   }
+
+   if (!ijmatrix)
+   {
+      hypre_error_in_arg(1);
+      return hypre_error_flag;
+   }
+
+   if (!ncols)
```
Comment (victorapm) [inline src/IJ_mv/HYPRE_IJMatrix.c:906]: In `HYPRE_IJMatrixSetValues2` we allow this pointer to be NULL. I'm wondering if it should be the same here

File: src/IJ_mv/HYPRE_IJMatrix.c [hunk 2]
```diff
@@ -814,7 +867,133 @@ HYPRE_IJMatrixGetValues( HYPRE_IJMatrix matrix,
...
```
Comment (victorapm) [inline src/IJ_mv/HYPRE_IJMatrix.c:980]: Same comment as for `HYPRE_IJMatrixGetValues2`

File: src/IJ_mv/HYPRE_IJ_mv.h [hunk 3]
```diff
@@ -222,6 +222,37 @@ HYPRE_Int HYPRE_IJMatrixGetValues(HYPRE_IJMatrix  matrix,
...
+                                   HYPRE_Complex  *values);
+
+/**
+ * Gets values for \e nrows rows or partial rows of the matrix
+ * and zeros out those entries in the matrix.
+ *
+ * Same as IJMatrixGetValues2, but zeros out the entries after getting them.
+ *
+ **/
+HYPRE_Int HYPRE_IJMatrixGetValuesAndZero(HYPRE_IJMatrix  matrix,
```
Comment (victorapm) [inline src/IJ_mv/HYPRE_IJ_mv.h]: Minor point: should we call this `HYPRE_IJMatrixGetValuesAndZeroOut`?

File: src/IJ_mv/IJMatrix_parcsr.c [hunk 4]
```diff
@@ -473,6 +475,7 @@ hypre_IJMatrixGetValuesParCSR( hypre_IJMatrix *matrix,
       col_map_offd = hypre_ParCSRMatrixColMapOffd(par_matrix);
    }
 
+   /* WM: question - what does rows < 0 signify? Do I need to add row_indexes here? */
```
Comment (victorapm) [inline src/IJ_mv/IJMatrix_parcsr.c]: @ulrikeyang I'm curious about this as well. What is the use case for `nrows < 0` here?

File: src/IJ_mv/_hypre_IJ_mv.h [hunk 5]
```diff
@@ -440,8 +440,8 @@ HYPRE_Int hypre_IJMatrixSetMaxOffProcElmtsParCSR ( hypre_IJMatrix *matrix,
 HYPRE_Int hypre_IJMatrixInitializeParCSR ( hypre_IJMatrix *matrix );
 HYPRE_Int hypre_IJMatrixGetRowCountsParCSR ( hypre_IJMatrix *matrix, HYPRE_Int nrows,
                                              HYPRE_BigInt *rows, HYPRE_Int *ncols );
-HYPRE_Int hypre_IJMatrixGetValuesParCSR ( hypre_IJMatrix *matrix, HYPRE_Int nrows, HYPRE_Int *ncols,
-                                          HYPRE_BigInt *rows, HYPRE_BigInt *cols, HYPRE_Complex *values );
+HYPRE_Int hypre_IJMatrixGetValuesParCSR ( hypre_IJMatrix *matrix, HYPRE_Int nrows, HYPRE_Int *ncols, HYPRE_BigInt *rows,
```
Comment (victorapm) [inline src/IJ_mv/_hypre_IJ_mv.h:443]: Should we add a warning that this does not support GPUs for now?

**Review by victorapm — APPROVED** [review]
Thank you, Wayne! I left some minor comments

File: src/IJ_mv/IJMatrix_parcsr.c (hunk 4 above)
Comment (rfalgout) [reply src/IJ_mv/IJMatrix_parcsr.c]: I'm pretty sure this means "get all rows".  I think we've also used something similar to "get all cols" for a given row.  Looking through the code here, that's what seems to be happening.

**Review by rfalgout — APPROVED** [review]
Looks good to me.  Thanks!

File: src/IJ_mv/IJMatrix_parcsr.c (hunk 4 above)
Comment (victorapm) [reply src/IJ_mv/IJMatrix_parcsr.c]: Thanks, Rob! In that case, it seems `row_indexes` is not needed here

File: src/IJ_mv/HYPRE_IJMatrix.c (hunk 1 above)
Comment (waynemitchell) [reply src/IJ_mv/HYPRE_IJMatrix.c:906]: `ncols` is used in `hypre_IJMatrixGetValuesParCSR()` and seems required to me in this case.

File: src/IJ_mv/HYPRE_IJMatrix.c (hunk 2 above)
Comment (waynemitchell) [reply src/IJ_mv/HYPRE_IJMatrix.c:980]: Same as above, I think `ncols` is necessary here.

File: src/IJ_mv/HYPRE_IJ_mv.h (hunk 3 above)
Comment (waynemitchell) [reply src/IJ_mv/HYPRE_IJ_mv.h]: Sure, I like that name better.

File: src/IJ_mv/_hypre_IJ_mv.h (hunk 5 above)
Comment (waynemitchell) [reply src/IJ_mv/_hypre_IJ_mv.h:443]: Added some "not implemented" error messages. Take a look at the additions and see whether that is what you had in mind.

File: src/IJ_mv/HYPRE_IJMatrix.c (hunk 1 above)
Comment (victorapm) [reply src/IJ_mv/HYPRE_IJMatrix.c:906]: Sounds good, thanks for clarifying!

File: src/IJ_mv/_hypre_IJ_mv.h (hunk 5 above)
Comment (victorapm) [reply src/IJ_mv/_hypre_IJ_mv.h:443]: Great, thanks!

**Pull request merged**